internal/hcl/project_locator.go

```diff
@@ -1458,20 +1455,55 @@ func hasDefaultVarFileExtension(name string) bool {
 	return false
 }
 
-// fullExtension returns the full extension of a file, starting from the first
-// dot in the file name. This is used instead of the builtin filepath.Ext
-// function as the latter only returns the last extension in the file name. For
-// example filepath.Ext("file.tfvars.json") would return ".json" instead of
-// ".tfvars.json".
-func fullExtension(fileName string) string {
-	// Find the index of the first dot.
-	dotIndex := strings.Index(fileName, ".")
-	if dotIndex == -1 {
-		// No dot found, return empty string.
-		return ""
-	}
-	// Return the substring from the first dot to the end of the string.
-	return fileName[dotIndex:]
+// splitExtension splits the extension from a file name. It will return the file name
+// without the extension and the extension itself. If the file name does not have an
+// extension it will return the original file name and an empty string.
+func splitExtension(fileName string, extensions []string) (string, string) {
+	if len(fileName) == 0 {
+		return "", ""
+	}
+
+	// Sort the extensions by length so that we can match the longest extension first.
+	sorted := make([]string, len(extensions))
+	copy(sorted, extensions)
+	sort.Slice(sorted, func(i, j int) bool {
+		return len(sorted[i]) > len(sorted[j])
+	})
+
+	for _, ext := range sorted {
+		if strings.HasSuffix(fileName, ext) {
+			return fileName[:len(fileName)-len(ext)], ext
+		}
+	}
+
+	return fileName, ""
+}
+
+// hasVarFileExtension checks if the file name has a valid extension.
+func hasVarFileExtension(fileName string, extensions []string) bool {
+	if len(fileName) == 0 {
+		return false
+	}
+
+	_, ext := splitExtension(fileName, extensions)
+	if ext != "" {
+		return true
+	}
+
+	blankExtensionAllowed := false
+	for _, e := range extensions {
+		if e == "" {
+			blankExtensionAllowed = true
+			break
+		}
+	}
+
+	// Check if there's a dot in the filename, but ignore the first character as we don't want to match hidden files.
+	if ext == "" && blankExtensionAllowed && !strings.Contains(fileName[1:], ".") {
+		return true
+	}
```

Comment:
Yeah this is a bit confusing. If blank extensions are allowed then `ext` being empty is valid as long as the filename doesn't have any other extension. If I rename `splitExtension` to `splitVarFileExt` and `ext` to `varFileExt` it might make it clearer. I'll also add some more docs around it.

`filepath.Ext` will return an ext for a `.hidden` file. I could replace `!strings.Contains(fileName[1:], ".")` `filepath.Ext(fileName[1:]) != ""`. I don't have any strong preference either way.